core/trino-main/src/test/java/io/trino/operator/TestDeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,443 @@
+            assertFalse(blocked.isDone());
+
+            TaskId taskId = createTaskId(0, 0);
+            buffer.addTask(taskId);
+            assertFalse(blocked.isDone());
+
+            buffer.taskFinished(taskId);
+            assertFalse(blocked.isDone());
+
+            buffer.noMoreTasks();
+            assertTrue(blocked.isDone());
+        }
+
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            ListenableFuture<Void> blocked = buffer.isBlocked();
+            assertFalse(blocked.isDone());
+
+            TaskId taskId = createTaskId(0, 0);
+            buffer.addTask(taskId);
+            assertFalse(blocked.isDone());
+
+            buffer.noMoreTasks();
+            assertFalse(blocked.isDone());
+
+            buffer.taskFinished(taskId);
+            assertTrue(blocked.isDone());
+        }
+
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            ListenableFuture<Void> blocked = buffer.isBlocked();
+            assertFalse(blocked.isDone());
+
+            TaskId taskId = createTaskId(0, 0);
+            buffer.addTask(taskId);
+            assertFalse(blocked.isDone());
+
+            buffer.taskFailed(taskId, new RuntimeException());
+            assertFalse(blocked.isDone());
+
+            buffer.noMoreTasks();
+            assertTrue(blocked.isDone());
+        }
+
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            ListenableFuture<Void> blocked = buffer.isBlocked();
+            assertFalse(blocked.isDone());
+
+            TaskId taskId = createTaskId(0, 0);
+            buffer.addTask(taskId);
+            assertFalse(blocked.isDone());
+
+            buffer.noMoreTasks();
+            assertFalse(blocked.isDone());
+
+            buffer.taskFailed(taskId, new RuntimeException());
+            assertTrue(blocked.isDone());
+        }
+
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            ListenableFuture<Void> blocked1 = buffer.isBlocked();
+            ListenableFuture<Void> blocked2 = buffer.isBlocked();
+            assertFalse(blocked1.isDone());
+            assertFalse(blocked2.isDone());
+
+            blocked2.cancel(true);
+
+            assertFalse(blocked1.isDone());
+            assertTrue(blocked2.isDone());
+        }
+    }
+
+    @Test
+    public void testPollPage()
+    {
+        testPollPages(ImmutableListMultimap.of(), ImmutableMap.of(), ImmutableList.of());
+        testPollPages(
+                ImmutableListMultimap.of(
+                        createTaskId(0, 0),
+                        createPage("p0a0v0")),
+                ImmutableMap.of(),
+                ImmutableList.of("p0a0v0"));
+        testPollPages(
```

Comment:
the limitation of this test method is that it always first registers all the tasks, before starting to add any pages. Maybe it does not matter given test coverage provide by other test methods here.

Reply:
Yeah, that was the assumption. This method is there to test the de-duplication algorithm (that the right pages are returned after de-duplication). Other tests also try to test different sequencing of events.